### ISSUE TITLE
[feature]: add `is_keysend` to `lnrpc.ListPaymentsRequest` and `lnrpc.ListInvoiceRequest`

### ISSUE THREAD

**Issue by secondl1ght** [lightningnetwork/lnd#7849]
**Is your feature request related to a problem? Please describe.**
For certain use cases you may only be interested in fetching the keysend payments that have been sent and received from a node. Currently this is not possible. You first need to fetch all of the data and then filter out the keysends by using `is_keysend` on the `Invoice` type. For the `Payment` type since there is no `is_keysend` I think you need to use `DecodePayReq` and then look at `Features` for `custom_records` to see if it is a keysend. Unless there is another way to tell? Related #7822.

**Describe the solution you'd like**
A simple param you can pass into these requests to only return keysends.

**Describe alternatives you've considered**
Filtering out the data client side - leads to a lot of extra data usage and CPU power.

**Additional context**

**Comment by alexbosworth** [comment]
if it's a keysend there wouldn't be a payreq, custom records are shown in received htlcs

**Comment by secondl1ght** [comment]
@alexbosworth oh right, I double checked the `Payment` type and there is no `Features` key there. So which endpoint can be used to list keysends that were made FROM the node including the custom records? I think `ListInvoice` only shows keysends (and other types) that were RECEIVED correct?

**Comment by alexbosworth** [comment]
You would see the custom records on the HTLCs and that indicates keysend

**Comment by secondl1ght** [comment]
Ok so just to confirm I can use `ListInvoices` to fetch both messages that were received AND sent? I've updated the issue, I still think this fetch param would be useful for this endpoint.

**Comment by alexbosworth** [comment]
list invoices are only received payments

**Comment by secondl1ght** [comment]
So it is not possible to view keysend custom records made from your node them? :\

**Comment by guggero** [comment]
It is, you just look in `ListPayments` instead.

**Comment by alexbosworth** [comment]
You can see those on the payment records, under the HTLCs, each HTLC has its own records

**Comment by secondl1ght** [comment]
Ok sorry that was confusing, I thought `ListPayments` is where I would find that information. Yes I see it now under `lnrpc.Hop.CustomRecordsEntry`, I updated the issue title again.

**Comment by secondl1ght** [comment]
Since the HTLCs is an array are you supposed to use any of them or the most recent one in order to look at the custom records? The custom records shouldn't change between HTLCs right?

**Comment by alexbosworth** [comment]
you can look at the successful one

lnd doesn't support keysend receiving on more than one htlc, although on sending potentially other implementations support multi path keysend

**Comment by secondl1ght** [comment]
Would it make sense to move the `custom_records` up to the top level of the `Payment` and `Invoice` types? If you set those at the initial creation and they can't change, then why do they need to be down many levels inside nested messages?

**Comment by secondl1ght** [comment]
Ah ok, but the custom records set by the sender are immutable? So the changes would be from the router?

**Comment by alexbosworth** [comment]
Yeah you don't update the custom records, per HTLC, but you could have multiple HTLCs and also failed and successful HTLCs

**Comment by secondl1ght** [comment]
But thats why I think that `custom_records` could be a key on `lnrpc.Invoice` or `lnrpc.Payment` because it sounds like that data is set on creation, so each HTLC would contain the same custom records so you could access that information higher in the Object. But maybe I am missing something... 

Anyway this issue and #7822 are about trying to find the most efficient way to reduce data load and client side filtering for fetching and accessing the custom record values on keysend payments. There are apps being built using these custom records that do not require any other type of transaction from the node, so I think having some optimisation for these use-cases would be great if possible. App builders can use the current tools and then adopt any improvements if they get released in the future.

Thanks for explaining how this works, now I know where to find the data needed. :)

**Comment by daveajones** [comment]
For what it’s worth, I agree with the sentiment of having keysend be a first class citizen in the LND API. In the absence of a widely adopted successor, we are stuck with Keysend as the only way to make spontaneous payments across the network with a decent chance of settling.  The Podcasting 2.0 ecosystem is fully reliant on this feature (again, until a better protocol is widely adopted) and has a high enough traffic volume that API efficiency does matter:  https://stats.podcastindex.org/v4v
